instrumentation/instaredis/instaredis_test.go

```diff
@@ -43,31 +45,77 @@ func (p *mockPipeline) Do(ctx context.Context, args ...interface{}) *redis.Cmd {
 	return cmd
 }
 
-func (p mockPipeline) Exec(ctx context.Context) ([]redis.Cmder, error) {
-	for _, hook := range p.hooks {
-		hook.BeforeProcessPipeline(ctx, p.cmds)
-		hook.AfterProcessPipeline(ctx, p.cmds)
+func (p *mockPipeline) Exec(ctx context.Context) {
+	p.process(ctx, p.cmds)
+}
+
+func (p *mockPipeline) process(ctx context.Context, cmds []redis.Cmder) (net.Conn, error) {
+	switch ctx.Value("pipe_type") {
+	case "processPipelineHook":
+		return nil, p.current.pipeline(ctx, cmds)
+	case "processTxPipelineHook":
+		return nil, p.current.txPipeline(ctx, cmds)
 	}
-	return p.cmds, nil
+	return nil, errors.New("unidentified pipe type")
 }
 
 type mockClient struct {
 	hooks   []redis.Hook
 	options *redis.Options
+	current hooks
+}
+
+type hooks struct {
+	dial       redis.DialHook
+	process    redis.ProcessHook
+	pipeline   redis.ProcessPipelineHook
+	txPipeline redis.ProcessPipelineHook
+}
+
+func newHooks() hooks {
+	return hooks{
+		dial:       func(ctx context.Context, network, addr string) (net.Conn, error) { return nil, nil },
+		process:    func(ctx context.Context, cmd redis.Cmder) error { return nil },
+		pipeline:   func(ctx context.Context, cmds []redis.Cmder) error { return nil },
+		txPipeline: func(ctx context.Context, cmds []redis.Cmder) error { return nil },
+	}
+}
+
+func (c *mockClient) chain() {
+	for i := len(c.hooks) - 1; i >= 0; i-- {
+		if wrapped := c.hooks[i].DialHook(c.current.dial); wrapped != nil {
```

Comment:
Are these ifs really necessary?
From what I see, `wrapped` will never be `nil`, but maybe I am missing something.
I also ran a quick test locally with this branch without the conditions and tests still pass normally.

If they are needed, could you kindly explain what is the logic behind it?

Reply:
Since we are mocking the client, I thought this would give a better understanding of the change in the underlying implementation of `go-redis` and this is taken from the original implementation. From a mock redis client perspective, it doesn't have to chain the middleware if it is returning `nil` for any of the implementation interface methods.